metricbeat/module/windows/services/services_windows.go

```diff
@@ -0,0 +1,355 @@
+	var processCreationTime gosigar.ProcTime
+
+	err := processCreationTime.Get(int(processID))
+	if err != nil {
+		return time.Duration(processCreationTime.StartTime), err
+	}
+
+	uptime := time.Since(time.Unix(0, int64(processCreationTime.StartTime)*int64(time.Millisecond))) / time.Millisecond
+
+	return uptime, nil
+}
+
+func getDetailedServiceInfo(handle ServiceDatabaseHandle, serviceName string, accessRight ServiceAccessRight, service *ServiceStatus) error {
+	var serviceBufSize uint32
+	var serviceBytesNeeded uint32
+
+	serviceHandle, err := OpenService(handle, service.ServiceName, ServiceQueryConfig)
+	if err != nil {
+		return err
+	}
+
+	if err := _QueryServiceConfig(serviceHandle, nil, serviceBufSize, &serviceBytesNeeded); err != nil {
+		if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_INSUFFICIENT_BUFFER {
+			if err := CloseServiceHandle(serviceHandle); err != nil {
+				return err
+			}
+			return err
+		}
+		serviceBufSize += serviceBytesNeeded
+		buffer := make([]byte, serviceBufSize)
+
+		for {
+			if err := _QueryServiceConfig(serviceHandle, &buffer[0], serviceBufSize, &serviceBytesNeeded); err != nil {
+				if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_INSUFFICIENT_BUFFER {
+					if err := CloseServiceHandle(serviceHandle); err != nil {
+						return err
+					}
+					return err
+				}
+				serviceBufSize += serviceBytesNeeded
+			} else {
+				serviceQueryConfig := (*QueryServiceConfig)(unsafe.Pointer(&buffer[0]))
+				service.StartType = serviceStartTypes[ServiceStartType(serviceQueryConfig.DwStartType)]
+				if err := CloseServiceHandle(serviceHandle); err != nil {
+					return err
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func (reader *ServiceReader) Close() error {
+	return CloseServiceDatabaseHandle(reader.handle)
+}
+
+func CloseServiceDatabaseHandle(handle ServiceDatabaseHandle) error {
+	if err := _CloseServiceHandle(uintptr(handle)); err != nil {
+		return ServiceErrno(err.(syscall.Errno))
+	}
+
+	return nil
+}
+
+func CloseServiceHandle(handle ServiceHandle) error {
+	if err := _CloseServiceHandle(uintptr(handle)); err != nil {
+		return ServiceErrno(err.(syscall.Errno))
+	}
+
+	return nil
+}
+
+func NewServiceReader() (*ServiceReader, error) {
+	hndl, err := OpenSCManager("", "", ScManagerEnumerateService|ScManagerConnect)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "initialization failed")
+	}
+
+	r := &ServiceReader{
+		handle: hndl,
+	}
+
+	r.state = ServiceStateAll
+
+	return r, nil
+}
+
+func (reader *ServiceReader) Read() ([]common.MapStr, error) {
+	services, err := getServiceStates(reader.handle, reader.state)
+
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]common.MapStr, 0, len(services))
+
+	for _, service := range services {
+		ev := common.MapStr{
+			"display_name": service.DisplayName,
+			"service_name": service.ServiceName,
+			"state":        service.CurrentState,
+			"start_type":   service.StartType,
+			"uptime": map[string]interface{}{
```

Comment:
If `service.Uptime` is 0 then let's not add it to the event. It should only be zero if an error occurred which would mean the data is invalid, right?

Reply:
Or the service is stopped. So maybe a check for `0` should help.